ToolBox/CoinApi/CoinApiDataQueueHandler.cs

```diff
@@ -0,0 +1,360 @@
+        public void Unsubscribe(LiveNodePacket job, IEnumerable<Symbol> symbols)
+        {
+            lock (_lockerSubscriptions)
+            {
+                var symbolsToUnsubscribe = (from symbol in symbols
+                                            where _subscribedSymbols.Contains(symbol)
+                                            select symbol).ToList();
+                if (symbolsToUnsubscribe.Count == 0)
+                    return;
+
+                Log.Trace($"CoinApiDataQueueHandler.Unsubscribe(): {string.Join(",", symbolsToUnsubscribe.Select(x => x.Value))}");
+
+                // CoinAPI requires at least 5 seconds between subscription requests so we need to batch them
+                _subscribedSymbols = _subscribedSymbols.Where(x => !symbolsToUnsubscribe.Contains(x)).ToHashSet();
+
+                ProcessSubscriptionRequest();
+            }
+        }
+
+        /// <summary>
+        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
+        /// </summary>
+        public void Dispose()
+        {
+            _connectionHandler.DisposeSafely();
+
+            if (_webSocket.IsOpen)
+            {
+                _webSocket.Close();
+            }
+        }
+
+        /// <summary>
+        /// Helper method used in QC backend
+        /// </summary>
+        /// <param name="markets">List of LEAN markets (exchanges) to subscribe</param>
+        public void SubscribeMarkets(List<string> markets)
+        {
+            if (markets.Count == 0) return;
+
+            Log.Trace($"CoinApiDataQueueHandler.SubscribeMarkets(): {string.Join(",", markets)}");
+
+            _subscribedExchanges = markets.ToList();
+
+            var message = JsonConvert.SerializeObject(new HelloMessage
+            {
+                ApiKey = _apiKey,
+                Heartbeat = true,
+                SubscribeDataType = new[] { "trade", "quote" },
+                SubscribeFilterSymbolId = markets.Select(x => _symbolMapper.GetExchangeId(x)).ToArray()
+            });
+
+            _webSocket.Send(message);
+
+            _connectionHandler.EnableMonitoring(true);
+        }
+
+        private void ProcessSubscriptionRequest()
+        {
+            if (_subscriptionsPending) return;
+
+            _lastSubscribeRequestUtcTime = DateTime.UtcNow;
+            _subscriptionsPending = true;
+
+            Task.Run(() =>
+            {
+                while (true)
+                {
+                    DateTime requestTime;
+                    List<Symbol> symbolsToSubscribe;
+                    lock (_lockerSubscriptions)
+                    {
+                        requestTime = _lastSubscribeRequestUtcTime.Add(_subscribeDelay);
+                        symbolsToSubscribe = _subscribedSymbols.ToList();
+                    }
+
+                    if (DateTime.UtcNow > requestTime)
```

Comment:
Seems there will always be a `_subscribeDelay` delay for the first subscriptions in the task, due to lines 179 + 190? Maybe it could set `_lastSubscribeRequestUtcTime` only _after_ it finishes each subscription request, line 201, and initialized at `MinValue` (& removing line 179)?

Reply:
The delay is intentional and expected as explained in the previous comment.